fix(linter): fix rule config not being correctly applied

camc314 opened the pull request:
since `rules_for_override` was converted from a `FxHashSet` to ` FxHashMap`, the behaviour changed. rather than calling `.replace`, we call `insert` which updates the value, but does not update the key.

As per the docs:

> The key is not updated, though; this matters for types that can be == without being identical.

I've written a test for this to hopefully prevent a regression and validated the test fails on main.


fixes #11054

camc314 commented:
* **#11055** <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/11055?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 👈 <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/11055?utm_source=stack-comment-view-in-graphite" target="_blank">(View in Graphite)</a>
* `main`
<hr />

## How to use the Graphite Merge Queue

Add either label to this PR to merge it via the merge queue:
- <kbd>0-merge</kbd> - adds this PR to the back of the merge queue
- <kbd>hotfix</kbd> - for urgent hot fixes, skip the queue and merge this PR next

You must have a Graphite account in order to use the merge queue. Sign up using [this link](https://app.graphite.dev/invite/github/oxc-project?ref=merge-queue-instructions-in-stack-comment&prId=11055).

<h2></h2>
<sub>

An organization admin has enabled the [Graphite Merge Queue](https://graphite.dev/docs/graphite-merge-queue) in this repository.

</sub>

<sub>

Please do not merge from GitHub as this will restart CI on PRs being processed by the merge queue.

</sub>

<h2></h2>

This stack of pull requests is managed by <a href="https://graphite.dev?utm-source=stack-comment"><b>Graphite</b></a>. Learn more about <a href="https://stacking.dev/?utm_source=stack-comment">stacking</a>.
<!-- Current dependencies on/for this PR: -->

camchenry reviewed (approved):
Sorry about that. Very surprised this didn't cause any tests to fail.

camc314 commented:
### Merge activity

* **May 15, 8:11 AM EDT**: The merge label '0-merge' was detected. This PR will be added to the [Graphite merge queue](https://app.graphite.dev/merges?org=oxc-project&repo=oxc) once it meets the requirements.
* **May 15, 8:11 AM EDT**: `camc314` added this pull request to the [Graphite merge queue](https://app.graphite.dev/merges?org=oxc-project&repo=oxc).

camc314 committed:
fix(linter): fix rule config not being correctly applied (#11055)

since `rules_for_override` was converted from a `FxHashSet` to ` FxHashMap`, the behaviour changed. rather than calling `.replace`, we call `insert` which updates the value, but does not update the key.

As per the docs:

> The key is not updated, though; this matters for types that can be == without being identical.

I've written a test for this to hopefully prevent a regression and validated the test fails on main.

fixes #11054
